new Gitlab CML doc (#2176)
iterative/dvc.org: content/docs/cml/start-gitlab.md

# Using CML on GitLab

Here, we'll walk through a tutorial to start using CML on GitLab. For
simplicity, we'll show the demo in GitLab Actions, but these instructions are

> valid for all supported Continuous Integration (CI) systems (with exceptions as

Reviewer: I think we should remove the second sentence- I wanted a disclaimer like that on the project README, where I didn't want people to assume only GitHub was supported. Here, in the context of the docs and the title of this page, I think we can remove it and just get right to the insturctions.

Reviewer 2: btw, probably `GitLab Actions` is not the right term?

Reviewer: correct, it's GitLab CI and GitHub Actions

noted!).

1. Fork our
   [example project repository](https://gitlab.com/iterative.ai/example_cml).
   Click on Fork and select the namespace where you would like to keep the
   project.

   ![](/img/gitlab_fork_cml_project.png)

> 2. ⚠️ In GitLab, to use CML, you must create a variable called a repo_token

Reviewer: Looks like `repo_token` should be in code formatting for consistency

   whose value is a Personal Access Token. To do this:

>    a. Click on your Avatar in the upper right side and click on Edit Profile.

Reviewer: It looks like, for consistency, "Edit Profile" should be in quotes like the other buttons are?

> # Using CML on GitLab

Reviewer 2: is it the same document? a duplicate?

Author: Not sure what you are referring to.  Same as what.  I think I accidentally made two pull requests for the same thing.  This is the most up to date.  under branch `improve-cml-docs`.

Reviewer 2: I see two documents here https://github.com/iterative/dvc.org/pull/2176/files: 

content/docs/cml/start-gitlab.md
content/docs/cml/get-started-with-cml-on-gitlab.md

Here, we'll walk through a tutorial to start using CML on GitLab. For
simplicity, we'll show the demo in GitLab Actions, but these instructions are

> valid for all supported Continuous Integration(CI) systems (with exceptions as

Reviewer 3: Typo. Here's a suggestion you can commit directly from GH:
```suggestion
valid for all supported Continuous Integration (CI) systems (with exceptions as
```

noted!).

1. Fork our
   [example project repository](https://gitlab.com/iterative.ai/example_cml).

>    Click on Fork and select what namespace to fork the project to.

Reviewer 3: - When I go to https://gitlab.com/iterative.ai/example_cml I don't see any "Fork" link or button.
- Minor: probably no need to use word "fork" 3 times.

Reviewer 3: > I don't see any "Fork" link

Oh, I wasn't signed in. Maybe mention that? Minor

Author: - It's in the upper right hand side.  There's a pic the with bounding box to clarify.
- LOL.  True. Let me see if I can figure out how you did that change in here....

>    ![](/img/gitlab_fork_cml_project.png)

Reviewer 3: BTW the image is great now but I just realized you can `<< Collapse sidebar` and then you get this view, which may be better for readability? 

![image](https://user-images.githubusercontent.com/1477535/108443285-24633980-721e-11eb-9027-299da40c5215.png)

You can also zoom in with the browser (Ctrl + `+`) until most of the width is used.

   > 💡 If just starting to use GitLab, you may receive a warning at the top of
   > the page regarding SSH keys. You do not need to use SSH keys for the
   > tutorial so you can click: "Don't show me again."  

>    > [Click this link for more information on SSH keys.](https://docs.gitlab.com/ee/ssh/)

Reviewer 3: 💅 this looks strange in the rendered page for some reason:

![image](https://user-images.githubusercontent.com/1477535/108442701-1bbe3380-721d-11eb-8d1e-df63562ef6b1.png)

Maybe make the link shorter so it fits in the 3rd line.

Reviewer 3: But wait... I'm not sure we need this note at all 🙂 

Users should probably figure out their SSH key situation separately.

2. ⚠️ CML in GitLab works exactly as in GitHub except for a few conventions. In

>    Gitlab, to use CML, you must create a variable called a repo_token whose

Reviewer 3: Prob no need to compare with the Github process: people going into this page will likely skip that one.

Reviewer 3: Also, 💅 the 2nd instance of "Gitlab" needs capital `L` (for consistency).

   value is a Personal Access Token. To do this:

   a. Click on your Avatar in the upper right side and click on Edit Profile.

   b. Along the left side of the screen go to Access Tokens.

   c. In the "Name" field, type `repo_token` and check boxes to select `api`,
   `read_repository` and `write_repository`.

   d. Click on the "Create personal access token" button and copy the generated
   access token.

   ![](/img/personal_access_token.png)

>    e. Head back to your forked repository by clicking the Projects tab next to
>    the GitLab logo and select your forked project.

Reviewer 3: Minor copy edit:
```suggestion
   e. Head back to your fork by clicking the Projects tab next to the GitLab
   logo and select it.
```

   f. On the left hand side Navigate to **Settings** ➡ **CI/CD** ➡ **Varibles**.

>    ![](/img/ci_cd_navigation.png)

Reviewer 3: Again, resolution = 💯 but maybe zoom into that corner for this one? Something like:

![image](https://user-images.githubusercontent.com/1477535/108446690-24fece80-7224-11eb-8c42-34cb6d83ef1f.png)

> Maybe a bit taller, up to you

   f. Scroll to Variables and expand the field. Click "Add Variable". In the Key
   field, type `repo_token`. In the Value field, paste your Personal Access
   Token. Check the "Mask variable" box, uncheck "Protect variable", and then
   save the variable by clicking "Add variable" at the bottom of the dialog box.

>    > 💡The following steps can all be done in the GitLab browser interface.

Reviewer 3: ```suggestion
   > 💡 The following steps can all be done in the GitLab browser interface.
```

>    > 💡The following steps can all be done in the GitLab browser interface.
>    > However, to follow along the commands, we recommend cloning your fork to
>    > your local workstation

Reviewer 3: Copy edit:
```suggestion
   > 💡 The following steps can all be done in the GitLab website.
   > However, to follow along the steps, we recommend cloning your fork to
   > your local workstation.
```
> This change would need rewrapping the paragraph though. But you can always let Restyled do that.

3. Go back to your forked `example_cml` project. Copy the Clone with HTTPS as
   shown in the image below, and then in your terminal, type the following

>    command, replacing <user_name> with your own from GitLab.

Reviewer 2: <user_name> -> `<user_name>`

   ![](/img/gitlab_cml_clone.png)

   ```bash
   git clone https://gitlab.com/<user_name>/example_cml.git
   ```

> 4. Change directory to example_cml

Reviewer 2: example_cml -> `example_cml` (inline code)

> 4. Change directory to `example_cml`

Reviewer 3: 💅 
```suggestion
4. Change directory to `example_cml`.
```

   ```bash
   cd example_cml
   ```

> 5. To create a CML workflow, use Vim or your editor of choice to copy the

Reviewer 3: Minor:
```suggestion
5. To create a CML workflow, use your editor of choice to copy the
```

   following into a new file `.gitlab-ci.yml` and save.

   ```yaml
   stages:
     - cml_run

   cml:
     stage: cml_run
     image: dvcorg/cml-py3:latest
     script:
       - pip3 install -r requirements.txt
       - python train.py

       - cat metrics.txt >> report.md
       - cml-publish confusion_matrix.png --md >> report.md
       - cml-send-comment report.md
   ```

> 6. In your text editor of choice, open `train.py` and edit line 16 to
>    `depth = 5`.

Reviewer 3: Minor:
```suggestion
6. In your text editor, open `train.py` and edit line 16 to `depth = 5`.
```

7. Commit and push the changes using:

   ```bash
   git checkout -b experiment
   git add . && git commit -m "modify forest depth"
   git push origin experiment
   ```

8. Go back to GitLab in a Browser window and create a merge request.

   ![](/img/create_merge_request.png)

9. If you arrive at a New Merge Request screen that says it's merging into
   anything _other_ than your local repository, click on `Change branches` seen
   here.

   ![](/img/new_merge_request.png)

10. ⚠️ Change target branch to your local branch with your username.

> ![](/img/change_user_name.png)

Reviewer 2: indent the image?

11. Click on the "Compare branches and continue" button. Enter any additional
    comments you would like to put in the description and click the "Submit

>     merge request" button. Shortly, you should see a comment from github-actions

Reviewer 3: github-actions ?
